Refactor weight rule utils: add to Execution API schema and consolidate enum definition

sunank200 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->
Refactor weight rule utils across Airflow components:

**Changes:**
1. Consolidate WeightRule enum definition in airflow-core/configuration.py and add to Execution API schema
2. Remove weight_rule.py from task-sdk definitions.
3. Move DB_SAFE_MINIMUM, DB_SAFE_MAXIMUM constants and db_safe_priority function to BaseOperator



closes: [#53391](https://github.com/apache/airflow/issues/53391)
related: [#53385](https://github.com/apache/airflow/issues/53385)
related: [#51671](https://github.com/apache/airflow/issues/51671)
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

sunank200 committed:
Move weight_rule utilities to Task SDK and deprecate via utils __init__.py; add version compatibility for tests

sunank200 committed:
Move weight_rule utilities to Task SDK and deprecate via utils __init__.py; add version compatibility for tests

amoghrajesh commented on airflow-core/src/airflow/utils/__init__.py:
```diff
@@ -46,5 +46,11 @@ def __getattr__(name: str):
+        "WeightRule": "airflow.sdk.definitions.weight_rule.WeightRule",
+        "db_safe_priority": "airflow.sdk.definitions.weight_rule.db_safe_priority",
+        "DB_SAFE_MINIMUM": "airflow.sdk.definitions.weight_rule.DB_SAFE_MINIMUM",
+        "DB_SAFE_MAXIMUM": "airflow.sdk.definitions.weight_rule.DB_SAFE_MAXIMUM",
```
These are only used in task sdk and that too in BaseOperator only. Move the constants and the method into BaseOperator i'd say. No point of keeping a module. And also the tests.

sunank200 replied:
Moved it

amoghrajesh commented on airflow-core/src/airflow/utils/__init__.py:
```diff
@@ -46,5 +46,11 @@ def __getattr__(name: str):
         "XCOM_RETURN_KEY": "airflow.models.xcom.XCOM_RETURN_KEY",
     },
+    "weight_rule": {
+        "WeightRule": "airflow.sdk.definitions.weight_rule.WeightRule",
```
WeightRule is used in config + sdk.

This one could probably just be duplicated.

sunank200 replied:
So this would make more sense to just move to airflow-shared i think

sunank200 replied:
For things which are shared between core and sdk like weight_rule, I have duplicated it in both places and added a TODO to remove it once airflow-shared is available. Or should i leave it for now?

sunank200 replied:
I have removed this for now.

amoghrajesh reviewed (commented):
Some comments, needs some work

sunank200 committed:
Move DB_SAFE_MINIMUM, DB_SAFE_MAXIMUM constants and db_safe_priority function into BaseOperator

Move DB_SAFE_MINIMUM, DB_SAFE_MAXIMUM constants and db_safe_priority function into BaseOperator

sunank200 committed:
Move DB_SAFE_MINIMUM, DB_SAFE_MAXIMUM constants and db_safe_priority function into BaseOperator

Move DB_SAFE_MINIMUM, DB_SAFE_MAXIMUM constants and db_safe_priority function into BaseOperator

ashb commented on airflow-core/src/airflow/utils/weight_rule.py:
```diff
@@ -21,19 +21,8 @@
-    return max(DB_SAFE_MINIMUM, min(DB_SAFE_MAXIMUM, priority_weight))
-
 
+# TODO: This can be moved to airflow-shared once it is available
```
Similar with TriggerRule, this should be an enum defined in the extra ExecAPI spec, so we share it via the spec, not via sharing the code.

A useful rule of thumb for when we want to share via the OpenAPI spec: Would this be useful/needed in the future when we have the ability to define DAGs in another language? If so -> enum in spec is the way to go.

sunank200 replied:
Makes sense. I have defined enum in the extra ExecAPI spec now.

ashb commented on task-sdk/src/airflow/sdk/definitions/weight_rule.py:
```diff
@@ -0,0 +1,45 @@
+        return set(cls.__members__.values())
+
+    def __str__(self) -> str:
+        return self.value
```
Are any of these used inside TaskSDK?

sunank200 replied:
Yes it is used indirectly.

Also i have removed this file completely as it is part of `airflow.sdk.api.datamodels._generated` now

sunank200 committed:
Define enum in the extra ExecAPI spec

sunank200 committed:
Define enum in the extra ExecAPI spec

amoghrajesh commented on airflow-core/src/airflow/utils/weight_rule.py:
```diff
@@ -1,56 +0,0 @@
-#
```
Module members need to be deprecated, you need to add deprecation warning for these. Example: https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/utils/__init__.py#L23-L47

sunank200 replied:
Deprecated them.

amoghrajesh commented on devel-common/src/tests_common/test_utils/compat.py:
```diff
@@ -63,7 +63,6 @@
     # Compatibility for Airflow < 3.1
     from airflow.utils.xcom import XCOM_RETURN_KEY  # type: ignore[no-redef,attr-defined]
 
-
```
We can remove this change

sunank200 replied:
Removed it

amoghrajesh replied:
nit: clear up this diff.

sunank200 replied:
Cleaned this

amoghrajesh commented on airflow-core/tests/unit/utils/test_weight_rule.py:
```diff
@@ -19,14 +19,7 @@
-from airflow.utils.weight_rule import DB_SAFE_MAXIMUM, DB_SAFE_MINIMUM, WeightRule, db_safe_priority
-
-
-def test_db_safe_priority():
-    assert db_safe_priority(1) == 1
-    assert db_safe_priority(-1) == -1
-    assert db_safe_priority(9999999999) == DB_SAFE_MAXIMUM
-    assert db_safe_priority(-9999999999) == DB_SAFE_MINIMUM
```
Do we not need this anymore?

sunank200 replied:
Not needed. Deleted it

amoghrajesh commented on airflow-core/tests/unit/utils/test_weight_rule.py:
```diff
@@ -19,14 +19,7 @@
 
 import pytest
 
-from airflow.utils.weight_rule import DB_SAFE_MAXIMUM, DB_SAFE_MINIMUM, WeightRule, db_safe_priority
```
All these tests can go in task sdk

sunank200 replied:
Moved them to sdk

amoghrajesh commented on airflow-core/src/airflow/configuration.py:
```diff
@@ -202,6 +203,28 @@ def retrieve_configuration_description(
+        return set(cls.__members__.values())
+
+    def __str__(self) -> str:
+        return self.value
```
This is only used in conf for `WeightRule.all_weight_rules()`, I wonder if we would be better off if we just hardcoded it.

cc @ashb / @kaxil wdyt?

sunank200 replied:
These methods can be removed. But this class is good to  have as we use it in task-sdk with execution API schema

amoghrajesh replied:
I'm ok with keeping it as well, although I really do not love that its defined in `airflow.configuration`, maybe we could move it elsewhere? I do not have a proposal at the moment

amoghrajesh commented on task-sdk/tests/test_bases/test_operator.py:
```diff
@@ -0,0 +1,34 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
```
test_operator doesn't seem right, are there no modules for operator already?

sunank200 replied:
Moved it to task-sdk/tests/task_sdk/bases/test_operator.py

sunank200 committed:
Add deprecation warnings for weight_rule module exports

amoghrajesh commented on task-sdk/src/airflow/sdk/definitions/_internal/abstractoperator.py:
```diff
@@ -31,13 +31,13 @@
 import methodtools
 
 from airflow.configuration import conf
+from airflow.sdk.api.datamodels._generated import WeightRule
```
I do not love this import pattern, IDK if its feasible to add it to lazy imports and have the usage be: `from airflow.sdk import WeightRule`?

sunank200 replied:
Changed it

amoghrajesh reviewed (approved):
I have some nits mainly, they aren't too pressing though.

sunank200 committed:
Add WeightRule to task-sdk lazy imports.

sunank200 reviewed (commented):
Resolved the comments

ashb commented on airflow-core/src/airflow/configuration.py:
```diff
@@ -202,6 +203,28 @@ def retrieve_configuration_description(
     return base_configuration_description
 
 
+class WeightRule(str, Enum):
```
What was you thinking for putting this in configuration.py? On seeing this it looks like a very odd location to put this enum

amoghrajesh replied:
It purely seems to be based on the usage here. Maybe a new module inside `airflow.operators` sounds better?

sunank200 replied:
It is only being used [configuration.py](https://github.com/astronomer/airflow/blob/fa038ddd2e92e0a304dd7fee2d8b55e7f126f7f6/airflow-core/src/airflow/configuration.py#L438) 

I did not find better place for this as it was not being used elsewhere

ashb replied:
https://github.com/astronomer/airflow/blob/fa038ddd2e92e0a304dd7fee2d8b55e7f126f7f6/airflow-core/src/airflow/models/taskinstance.py#L547-L549 -- that's where it's actually used in the server side of things.

ashb replied:
However I'm not sure if weight_rule has actually been working correctly since https://github.com/astronomer/airflow/commit/9855515becfb85a06dc9a64e50b0c5b4e5a6597c landed. Iiinteresting.

ashb replied:
Ah, and here is the other place it's used https://github.com/astronomer/airflow/blob/9c7c003465c2261132909a26b7262e44d9a0e1d5/airflow-core/src/airflow/task/priority_strategy.py#L113-L117

Those strings should have been the weight rule enum values. That's why no uses were showing up properly.

sunank200 replied:
But should this be part of this PR? I think `PriorityWeightStrategy` is already in task and not inside utils.

ashb replied:
Either way, putting this enum in configuration.py is not right -- it's nothing to do with config but dags/tasks

sunank200 replied:
I will move it to task as `PriorityWeightStrategy` is already in task.

sunank200 replied:
I am also using WeightRule class object in priority_strategy.py

amoghrajesh replied:
Yeah that would be nice to do.

amoghrajesh replied:
Resolving for now.

sunank200 committed:
Fix the test

sunank200 committed:
Fix the extra space

sunank200 committed:
Move weight_rule to airflow-core/src/airflow/task/weight_rule.py

amoghrajesh commented on airflow-core/src/airflow/task/weight_rule.py:
```diff
@@ -21,21 +21,9 @@
-    return max(DB_SAFE_MINIMUM, min(DB_SAFE_MAXIMUM, priority_weight))
-
 
 class WeightRule(str, Enum):
```
Yeah this move sounds like a better idea.

amoghrajesh commented on task-sdk/tests/task_sdk/docs/test_docs_inventory.py:
```diff
@@ -76,7 +76,7 @@ def test_docs_inventory_matches_public_api(tmp_path):
 
     extras = {"AirflowParsingContext"}
-    missing = public - documented
+    missing = (public - documented) - {"WeightRule"}
```
We will have many such in near future I assume, so could you instead declare a `excluded` set with those?

sunank200 replied:
Changed it

amoghrajesh commented on airflow-core/src/airflow/task/weight_rule.py:
```diff
@@ -21,21 +21,9 @@
-    return max(DB_SAFE_MINIMUM, min(DB_SAFE_MAXIMUM, priority_weight))
-
 
 class WeightRule(str, Enum):
```
Could we also update: `airflow-core/src/airflow/task/priority_strategy.py` to use the values from this enum?

sunank200 replied:
Good idea. Changed it.

amoghrajesh reviewed (approved):
This looks better to me now, some nits.

sunank200 committed:
Use the WeightRule in task/priority_strategy.py
